jvmhs/src/Jvmhs/Analysis/Reduce.hs

```diff
@@ -92,3 +93,100 @@ findUsedClasses  =
           (traverse.classNames)
           (traverse.classNames)
           nothing) cls
+
+
+chopVector :: Int -> V.Vector x -> V.Vector (V.Vector x)
+chopVector n vec
+  | divisible = chopVector' blockSize n vec V.empty
+  | otherwise = chopVector' blockSize (vectorLength `mod` n) vec V.empty
+  where blockSize = vectorLength `roundUpQuot` n
+        vectorLength = length vec
+        divisible =  length vec `mod` n == 0
+
+chopVector' ::
+     Int
+  -- ^ size of each Vec (delta)
+  -> Int
+  -- ^ num of vec remaining should be larger
+  -> V.Vector x
+  -> V.Vector (V.Vector x)
+  -> V.Vector (V.Vector x)
+
+chopVector' i n vec rslt
+  | V.null vec        = rslt
+  | otherwise         = let (block, rest) = V.splitAt newBlockSize vec
+                          in chopVector'
+                            newBlockSize newN rest (snoc rslt block)
+  where newN         = if n > -1 then n - 1 else n
+        newBlockSize = if n == 0 then i-1 else i
+
+roundUpQuot :: Int -> Int -> Int
+roundUpQuot i j = ceiling ((fromIntegral i / fromIntegral j)::Float)
+
+deltaI ::
+     V.Vector (V.Vector x)
+  -> Int
+  -> V.Vector x
+deltaI vecs i =
+  vecs V.! i
+
+deltaComplementI ::
+    V.Vector (V.Vector x)
+  -> Int
+  -> V.Vector x
+deltaComplementI vecs i =
+  V.foldl1 (V.++) $ V.ifilter (\j _ -> i /= j) vecs
+
+
+ddmin ::
+     Monad m
+--     (Monad m, MonadIO m, Show x)
+  => V.Vector x
+  -> (V.Vector x -> m Bool)
+  -- ^ a function can test delta
+  -> m(V.Vector x)
+
+ddmin world = ddmin' world 2
+
+ddmin' ::
+     Monad m
+--     (Monad m, MonadIO m, Show x)
+  => V.Vector x
+  -> Int
+  -> (V.Vector x -> m Bool)
+  -- ^ a function can test delta
+  -> m(V.Vector x)
+ddmin' world n testFunc = do
+  deltaTrueVec <- V.filterM testFunc deltaVec
+  deltaComplTrueVec <- V.filterM testFunc deltaComplVec
+  if not $ null deltaTrueVec
```

Comment:
This looks like pattern matching, maybe you want to use a maybe instead of a vector, 
like this:
```haskell
case (deltaTrueVec , deltaComplTrueVect) of 
  (Just a , _      ) ->
    ddmin' a 2 testFunc
  (Nothing, Just a ) -> 
    ddmin' a (max (n-1) 2) testFunc
  (Nothing, Nothing) 
     | n < V.length world -> 
        ddmin' world (min (V.length world) (2*n)) testFunc
     | otherwise -> 
        return world
```

Reply:
Also: 
```haskell
     | n < V.length world -> 
        ddmin' world (min (V.length world) (2*n)) testFunc
```
We can propably change this to: 
```haskell
     | 2 * n < V.length world -> 
        ddmin' world (2*n) testFunc
```

Reply:
Why are they the same? What happen n > length and 2n < length?

Reply:
In that case, we would call `ddmin' world (V.length world) testFunc`, which would call the 
method again with size `n = (V.length world)`. Since we have already tested for n >= V.length world we can't cop the list down in smaller bits than 1 in each. 

Hm.. the only thing we need to make sure is the case where V.length world = n. Maybe I'm wrong :)